local jecs = require("@jecs")
local lifetime_tracker_add = require("@tools/lifetime_tracker")
local world = lifetime_tracker_add(jecs.world())
world:print_snapshot()
local e = world:entity()
local e1 = world:entity()
world:delete(e)

world:print_snapshot()
local e2 = world:entity()
local e3 = world:entity()
world:print_snapshot()
world:delete(e1)
world:delete(e2)
world:delete(e3)
world:print_snapshot()
world:print_entities()
world:entity()
world:entity()
world:print_snapshot()
